### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by vsupruniuk**

### CODE REVIEW

File: src/main/java/practice/CandidateValidator.java [hunk 1]
```diff
@@ -1,5 +1,27 @@
...
+    public boolean test(Candidate candidate) {
+        return candidate.getAge() >= MIN_AGE
+                && candidate.isAllowedToVote()
+                && candidate.getNationality().equals(ALLOWED_NATIONALITY)
+                && calculateYearsInUkraine(candidate.getPeriodsInUkr()) >= MIN_YEARS_IN_UKRAINE;
+    }
+
+    private int calculateYearsInUkraine(String years) {
+        return Arrays.stream(years
+                .split("-"))
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/CandidateValidator.java:22]: better move the hyphen to constant fields

File: src/main/java/practice/CandidateValidator.java [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
-public class CandidateValidator {
-    //write your code here
+import java.util.Arrays;
+import java.util.function.Predicate;
+import model.Candidate;
+
+public class CandidateValidator implements Predicate<Candidate> {
+    public static final String ALLOWED_NATIONALITY = "Ukrainian";
+    public static final int MIN_AGE = 35;
+    public static final int MIN_YEARS_IN_UKRAINE = 10;
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/CandidateValidator.java:10]: these constants can be private

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -50,7 +71,13 @@ public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toA
      */
     public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        return peopleList
+                .stream()
+                .filter(p -> p.getAge() >= fromAge
+                && ((p.getSex() == Person.Sex.MAN && p.getAge() <= maleToAge)
+                        || (p.getSex() == Person.Sex.WOMAN && p.getAge() <= femaleToAge))
+                )
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/StreamPractice.java:79]: Let’s  use Predicate for complicated conditionals

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -23,7 +34,12 @@ public int findMinEvenNumber(List<String> numbers) {
      * But before that subtract 1 from each element on an odd position (having the odd index).
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        return IntStream
+                .range(0, numbers.size())
+                .map(i -> i % 2 == 0 ? numbers.get(i) : numbers.get(i) - 1)
+                .filter(n -> n % 2 != 0)
+                .average()
+                .orElseThrow();
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/StreamPractice.java:42]: Let’s  throw some exception

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -14,7 +16,16 @@ public class StreamPractice {
      * "Can't get min value from list: < Here is our input 'numbers' >"
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers
+                .stream()
+                .flatMap(n -> Arrays.stream(n.split(",")))
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/StreamPractice.java:21]: better move coma to constant field ;)

File: src/main/java/practice/CandidateValidator.java [hunk 6]
```diff
@@ -1,5 +1,27 @@
...
+                && candidate.getNationality().equals(ALLOWED_NATIONALITY)
+                && calculateYearsInUkraine(candidate.getPeriodsInUkr()) >= MIN_YEARS_IN_UKRAINE;
+    }
+
+    private int calculateYearsInUkraine(String years) {
+        return Arrays.stream(years
+                .split("-"))
+                .mapToInt(Integer::parseInt)
+                .reduce((a, b) -> Math.abs(a - b))
+                .orElse(0);
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/CandidateValidator.java:25]: why 0? Maybe better throw some exceptions?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

File: src/main/java/practice/StreamPractice.java (hunk 4 above)
Comment (vsupruniuk) [reply src/main/java/practice/StreamPractice.java:42]: NoSuchElementException is a task requirement, but I'll add some text to the exception

File: src/main/java/practice/CandidateValidator.java (hunk 6 above)
Comment (vsupruniuk) [reply src/main/java/practice/CandidateValidator.java]: I thought about edge cases like "2020-2020". But nvm, I have a better idea about this solution

File: src/main/java/practice/CandidateValidator.java [hunk 7]
```diff
@@ -1,5 +1,43 @@
...
+    @Override
+    public boolean test(Candidate candidate) {
+        return candidate.getAge() >= MIN_AGE
+                && candidate.isAllowedToVote()
+                && candidate.getNationality().equals(ALLOWED_NATIONALITY)
+                && calculateYearsInUkraine(candidate.getPeriodsInUkr()) >= MIN_YEARS_IN_UKRAINE;
+    }
+
+    private int calculateYearsInUkraine(String years) {
+        IntBinaryOperator numbersDiffOrException = (a, b) -> {
```
Comment (liliia-ponomarenko) [inline src/main/java/practice/CandidateValidator.java:23]: okay) good practice with functional interfaces ;)

**Review by liliia-ponomarenko — APPROVED** [review]
Great 😊
